### PR TITLE
initwd Windows tests fixes

### PR DESCRIPTION
**Pull request by eduzgun**
## Target Release

1.8.0

## Checklist

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.

### Go checklist

 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [x] I have added tests for all relevant use cases of my code, and those tests are passing.
- [x] I have only exported functions, variables and structs that should be used from other packages.
- [x] I have added meaningful comments to all exported functions, variables, and structs.

Related to #1201 

This PR helps resolves the initwd package tests failing on Windows. The errors included Windows filepath errors and a Tempdir Cleanup RemoveAll error. The solution for TempDir RemoveAll Cleanup error was to do `runtime.GC()` which runs a garbage collection and blocks other processes. This worked because it most likely released the open file handles that were not needed anymore. In the `WriteSnapshotToDir` function I also closed the file after creating it.

## Important
Another test was failing which was `TestModuleInstaller_symlink` and I was getting this error:
```
--- FAIL: TestModuleInstaller_symlink (0.01s)
    module_install_test.go:377: failed to copy fixture to temporary directory: symlink ..\child_a\ C:\Users\ED\AppData\Local\Temp\terraform-configload2464638177\modules\child_a: A required privilege is not held by the client.
```
What I had to do is enable developer mode on Windows (11), see https://github.com/Schniz/fnm/issues/338. If it's not possible to enable developer mode in the production environment then we could skip this test for Windows runtime only.
```
if runtime.GOOS == "windows" {
    t.Skip("Skipping symlinks test on Windows")
}
```

Tests in this package are passing on Windows and Linux machine
initwd pass

### CODE REVIEW

File: internal/initwd/from_module_test.go
Comment (cam72cam) [inline internal/initwd/from_module_test.go:285]: Can you explain / add a comment why we need manual GC here?

File: internal/initwd/module_install_test.go
Comment (cam72cam) [inline internal/initwd/module_install_test.go:746]: This is very strange, I wonder how `.` is being replaced with `\` in the module install hook

File: internal/initwd/from_module_test.go
Comment (eduzgun) [reply internal/initwd/from_module_test.go:287]: Sure, there are many TempDir RemoveAll cleanup errors that are coming up when running tests on Windows. This is because a file is being written to or created without properly being closed. As a result, TempDir cannot delete the temp files on Windows because they are blocked by another process. `t.cleanup()` is run when the test completes so it doesn't not affect the logic inside the test. By manually calling `runtime.GC()`, it triggers finalisers to run. If a file has no more references and a finaliser is set to close the file then `runtime.GC()` will cause this file handle, that is no longer being used, to close. Then the temporary file in Windows can be deleted.

File: internal/initwd/module_install_test.go
Comment (eduzgun) [reply internal/initwd/module_install_test.go:746]: I agree, I got stuck there looking in the code. Myabe it's the way the testdata`testdata/local-module-from-test` is laid out?

File: internal/initwd/from_module_test.go
Comment (cam72cam) [reply internal/initwd/from_module_test.go:287]: Thanks!

File: internal/initwd/module_install_test.go
Comment (cam72cam) [reply internal/initwd/module_install_test.go:746]: Honestly, this seems like it might actually be a bug.  I'm digging into it a bit further.

Comment (cam72cam) [reply internal/initwd/module_install_test.go:746]: I'm looking at ModuleInstaller.installLocalModule and moduleInstallWalker, which leads me to configs.ModuleRequest.Path (joined with '.').

internal/configs/config_build.go:buildTestModules is probably where the req.Path is set and includes some logic around splitting on "/".  I'm guessing that's where the bug is.

Comment (eduzgun) [reply internal/initwd/module_install_test.go]: Yes you were correct. I've added a commit to this PR which uses filepath instead of path to solve that issue. Did a little test to show that `path.base()` always assumes the file system is UNIX:
```
var a_path_linux string = "bananas/main.tftest.hcl"
var a_path_windows string = "bananas\\main.tftest.hcl"

fmt.Println("The linux:", path.Base(a_path_linux))
fmt.Println("The windows: ", path.Base(a_path_windows))

fmt.Println("The linux filepath base:", filepath.Base(a_path_linux))
fmt.Println("The windows filepath base: ", filepath.Base(a_path_windows))

Output:
The linux: main.tftest.hcl
The windows:  bananas\main.tftest.hcl
The linux filepath base: main.tftest.hcl
The windows filepath base:  main.tftest.hcl
```

**Comment by eduzgun** [comment]
Also, is it a better idea to do 
```
t.Cleanup(func() {
    if runtime.GOOS == "windows" {
        runtime.GC()
    }
})
```
so we only do garbage collection on Windows? I'm assuming these tests were already passing on Linux. If so, we don't have to touch manual garbage collection for Linux.

**Comment by eduzgun** [comment]
Okay, I've updated it now. Thanks for helping

**Comment by cam72cam** [comment]
@eduzgun it looks like your latest commit is missing the DCO signoff

**Review by cam72cam — APPROVED** [review]

**Comment by cam72cam** [comment]
@eduzgun looks like one minor linting issue and then we are good to merge :+1:

**Pull request merged**
